flink-core/src/main/resources/PojoSerializerTemplate.ftl

```diff
@@ -0,0 +1,372 @@
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.typeutils.TypeExtractor;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+
+public final class ${className} extends TypeSerializer {
+	private static byte IS_NULL = 1;
+	private static byte NO_SUBCLASS = 2;
+	private static byte IS_SUBCLASS = 4;
+	private static byte IS_TAGGED_SUBCLASS = 8;
+	private int numFields;
+	private ExecutionConfig executionConfig;
+	private Map<Class, TypeSerializer> subclassSerializerCache;
+	private final Map<Class, Integer> registeredClasses;
+	private final TypeSerializer[] registeredSerializers;
+	Class clazz;
+	<#list members as m>
+	${m}
+	</#list>
+	public ${className}(Class clazz, TypeSerializer[] serializerFields, ExecutionConfig e) {
+		this.clazz = clazz;
+		executionConfig = e;
+		this.numFields = serializerFields.length;
+		LinkedHashSet<Class> registeredPojoTypes = executionConfig.getRegisteredPojoTypes();
+		subclassSerializerCache = new HashMap<Class, TypeSerializer>();
+		List<Class> cleanedTaggedClasses = new ArrayList<Class>(registeredPojoTypes.size());
+		for (Class registeredClass: registeredPojoTypes) {
+			if (registeredClass.equals(clazz)) {
+				continue;
+			}
+			if (!clazz.isAssignableFrom(registeredClass)) {
+				continue;
+			}
+			cleanedTaggedClasses.add(registeredClass);
+		}
+		this.registeredClasses = new LinkedHashMap<Class, Integer>(cleanedTaggedClasses.size());
+		registeredSerializers = new TypeSerializer[cleanedTaggedClasses.size()];
+		int id = 0;
+		for (Class registeredClass: cleanedTaggedClasses) {
+			this.registeredClasses.put(registeredClass, id);
+			TypeInformation typeInfo = TypeExtractor.createTypeInfo(registeredClass);
+			registeredSerializers[id] = typeInfo.createSerializer(executionConfig);
+			id++;
+		}
+		<#list initMembers as m>
+		${m}
+		</#list>
+	}
+	private TypeSerializer getSubclassSerializer(Class subclass) {
+		TypeSerializer result = (TypeSerializer)subclassSerializerCache.get(subclass);
+		if (result == null) {
+			TypeInformation typeInfo = TypeExtractor.createTypeInfo(subclass);
+			result = typeInfo.createSerializer(executionConfig);
+			subclassSerializerCache.put(subclass, result);
+		}
+		return result;
+	}
+	public boolean isImmutableType() { return false; }
+	public ${className} duplicate() {
+		boolean stateful = false;
+		TypeSerializer[] duplicateFieldSerializers = new TypeSerializer[numFields];
+		<#list duplicateSerializers as ds>
+		${ds}
+		</#list>
+		if (stateful) {
+			return new ${className}(clazz, duplicateFieldSerializers, executionConfig);
+		} else {
+			return this;
+		}
+	}
+	public ${typeName} createInstance() {
+		if (clazz.isInterface() || Modifier.isAbstract(clazz.getModifiers())) {
+			return null;
+		}
+		try {
+			${typeName} t = (${typeName})clazz.newInstance();
```

Comment:
You can generate `new ${typeName}` here, which avoids reflection. Also, the above checks can be performed at code generation time.